### ISSUE TITLE
keras.MultiHeadAttention fails to save / load weights correctly

### ISSUE THREAD

**Issue by melton1968** [tensorflow/tensorflow#47722]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):

Yes,
```
import tensorflow.keras as tk
import numpy as np

def main(**kwargs):
    inputs = tk.Input(shape=(1,2))
    outputs = tk.layers.MultiHeadAttention(1, 2)(inputs, inputs)
    model = tk.Model(inputs, outputs)
    model.compile(tk.optimizers.Adam(),'MSE')

    #Adding .fit call doesn't change the result
    #data = np.ones((1,1,2))
    #model.fit(x=data, y=data)

    model.save("bug0")
    m0 = tk.models.load_model("bug0")
    m1 = tk.models.load_model("bug0")

    print()
    print(m0.weights[0])
    print(m1.weights[0])


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(e)
```
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):

```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.2 LTS
Release:        20.04
Codename:       focal

```
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:n/a
- TensorFlow installed from (source or binary):

```
pip install tensorflow
```

- TensorFlow version (use command below):
```
$ python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"
v2.4.0-49-g85c8b2a817f 2.4.1
```
- Python version:
```
$ python --version
Python 3.8.5
```
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

When loading a saved model containing a MultiHeadAttention layer, the layer wights are different every time the model is loaded.

**Describe the expected behavior**

The layer weights should be the same every time the model is loaded.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```
import tensorflow.keras as tk
import numpy as np

def main(**kwargs):
    inputs = tk.Input(shape=(1,2))
    outputs = tk.layers.MultiHeadAttention(1, 2)(inputs, inputs)
    model = tk.Model(inputs, outputs)
    model.compile(tk.optimizers.Adam(),'MSE')

    #Adding .fit call doesn't change the result
    #data = np.ones((1,1,2))
    #model.fit(x=data, y=data)

    model.save("bug0")
    m0 = tk.models.load_model("bug0")
    m1 = tk.models.load_model("bug0")

    print()
    print(m0.weights[0])
    print(m1.weights[0])


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(e)
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

```
2021-03-11 07:17:30.470279: W tensorflow/python/util/util.cc:348] Sets are not currently considered\
 sequences, but this may change in the future, so consider avoiding using them.
WARNING:absl:Found untraced functions such as query_layer_call_fn, query_layer_call_and_return_cond\
itional_losses, key_layer_call_fn, key_layer_call_and_return_conditional_losses, value_layer_call_f\
n while saving (showing 5 of 30). These functions will not be directly callable after loading.
WARNING:absl:Found untraced functions such as query_layer_call_fn, query_layer_call_and_return_cond\
itional_losses, key_layer_call_fn, key_layer_call_and_return_conditional_losses, value_layer_call_f\
n while saving (showing 5 of 30). These functions will not be directly callable after loading.

<tf.Variable 'multi_head_attention/query/kernel:0' shape=(2, 1, 2) dtype=float32, numpy=
array([[[ 0.36944628, -0.00638223]],

       [[-0.96410084, -0.4008801 ]]], dtype=float32)>
<tf.Variable 'multi_head_attention/query/kernel:0' shape=(2, 1, 2) dtype=float32, numpy=
array([[[ 0.94459724,  0.45157957]],

       [[ 0.88224053, -0.97082996]]], dtype=float32)>
```

**Comment by melton1968** [comment]
Manually saving and loading the weights works as expected (still get a lot of warnings). The `tf-nightly` version produces the same results except only one warning instead of two.

```
model.save("bug0")
model.save_weights("bug0-w")

m0 = tk.models.load_model("bug0")
m0.load_weights("bug0-w")

m1 = tk.models.load_model("bug0")
m1.load_weights("bug0-w")
```

```
WARNING:absl:Found untraced functions such as query_layer_call_and_return_conditional_losses, query\
_layer_call_fn, key_layer_call_and_return_conditional_losses, key_layer_call_fn, value_layer_call_a\
nd_return_conditional_losses while saving (showing 5 of 30). These functions will not be directly c\
allable after loading.
WARNING:absl:Found untraced functions such as query_layer_call_and_return_conditional_losses, query\
_layer_call_fn, key_layer_call_and_return_conditional_losses, key_layer_call_fn, value_layer_call_a\
nd_return_conditional_losses while saving (showing 5 of 30). These functions will not be directly c\
allable after loading.

<tf.Variable 'multi_head_attention/query/kernel:0' shape=(2, 1, 2) dtype=float32, numpy=
array([[[-0.37047267, -0.5616925 ]],

       [[ 0.6572478 , -0.18718529]]], dtype=float32)>
<tf.Variable 'multi_head_attention/query/kernel:0' shape=(2, 1, 2) dtype=float32, numpy=
array([[[-0.37047267, -0.5616925 ]],

       [[ 0.6572478 , -0.18718529]]], dtype=float32)>
```

**Comment by amahendrakar** [comment]
@melton1968,
I was able to save and load the weights without any issues by
- saving the model as a `.h5` file or
- using the `save_weights` and `load_weights` methods.

Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/9ecc22a8e58efff7e3e923351aa7a067/47722.ipynb#scrollTo=pWLoZ4L65xHR) and also take a look at similar issues [#47487](https://github.com/tensorflow/tensorflow/issues/47487) [#33424](https://github.com/tensorflow/tensorflow/issues/33424#issuecomment-544933871).

Regarding the warnings logs, the issue is already being tracked in issue [#47479](https://github.com/tensorflow/tensorflow/issues/47479).

Thanks!

**Comment by melton1968** [comment]
@amahendrakar,

My actual use-case involves a subclassed Model which is not supported by the hd5 format:

```
Saving the model to HDF5 format requires the model to be a Functional model or a Sequential model.  It does not work for subclassed models, because such models are defined via the body of a Python method, which isn't safely serializable. Consider saving to the Tensorflow SavedModel format (by setting save_format="tf") or using `save_weights`.
```

Using `save_weights` and `load_weights` does seem to save and load the weights correctly. Unfortunate that the standard `save/load` methods are broken.

**Comment by ymodak** [comment]
This is due to random values for `kernel_initializer` in [MultiHeadAttentionLayer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/MultiHeadAttention)
Setting seed for `kernel_initializer= tk.initializers.glorot_uniform(seed=0)` can help to load same weights in your case.
Refer [gist](https://colab.research.google.com/gist/ymodak/ed3d8d9e171a5fbf7a313d29cd20cdd7/47722.ipynb)
